### ISSUE TITLE
Add support for semgrep rules that check for specific file path names

### ISSUE THREAD

**Issue by rlundinSlack** [semgrep/semgrep#2438]
**Is your feature request related to a problem? Please describe.**
I am working on creating a custom semgrep ruleset and would like to be able to check the target repository for a specific filepath, directory name, or file name.  For example, in a golang repository, I would like to be able to check for the existence of a "vendor" directory.

**Describe the solution you'd like**
It would be great if I could add a field to the rule's yaml file that looks for a pattern that matches a file path rather than looking for a pattern within the file itself.  For example if I could add the following rule with a "pattern-filepath" field, my specific problem would be solved:

```
rules:
  - id: detect-vendor-directory
    message: "Detected a vendor directory"
    languages: [go]
    severity: INFO
    pattern-filepath: "vendor/"
```

**Describe alternatives you've considered**
@clintgibler mentioned the following existing workaround in the r2c Slack workspace that will work to detect filepaths for the time being.  This workaround helps but is not the most straightforward or simple solution.

```
rules:
  - id: find-go-mod-file
    message: "Detected a go.mod file."
    languages: [generic]
    severity: INFO
    pattern-regex: "^.*"
    paths:
      include:
        - "go.mod"
      exclude:
        - "vendor/"
```

**Additional context**
See r2c Slack workspace's "rules" channel for the original thread re: this feature request.

**Comment by karoun** [comment]
This would be terrific, allowing semgrep to take the place of tools like `repolinter` or `ls-lint`.

**Comment by DrewDennison** [comment]
@mschwager this reminds me of the great work you did using regexs and the boolean logic for `pattern-regex` Do you think it would be feasible to wrap `find` or python's Pathlib in a similar fashion?

**Comment by Isaac0616** [comment]
+1 to this feature, it would be very helpful to my use cases.
In addition, I would love to be able to use regex to match the path.

**Comment by bertonjulian** [comment]
+1 also to this feature (unless there is another way to check this), as we currently use something like this for our rules and it would help to have a rule checking that this file still exists as we want to make sure our rules don't silently fail if devs change the name of the file we are searching in.

```  
paths:
    include:
      - src/apis/authRoutes.ts
```

**Comment by ievans** [comment]
I believe this is mostly solved now with logic under the ‘paths’ key

**Comment by mschwager** [comment]
Following up here, this is the official workaround we use internally:

```yaml
rules:
  - id: found-file
    message: Detected a file
    languages: [generic]
    severity: INFO
    pattern-regex: ^.*
    paths:
      include:
        - <filename-or-glob-here>
```

E.g.

```yaml
rules:
  - id: found-dockerfile
    message: Detected a Dockerfile
    languages: [generic]
    severity: INFO
    pattern-regex: ^.*
    paths:
      include:
        - Dockerfile
```

**Comment by clintgibler** [comment]
Quick handwavy spec at what first class support for this might look like.

New clause:
* `pattern-file`: Accepts a file glob, like under the `paths` key. Returns any file that matches the path glob. 
  * The matched text could be empty and the range could just be starting line and ending line and column of (0,0) or something. Basically, the user should expect no content from the file back from this, just what files were matched.

Note: `pattern-file` *cannot* be combined with any other `pattern` clauses. If the user wants to match based on file contents, they should use the `paths` key. So if Semgrep detects `pattern-file` being used with other `pattern-foo` clauses, it should print an error message describing how the clauses should be used.

### Mode

For `mode`, we could re-use `generic` to not make users remember another thing, but it's different than `generic` in that Semgrep won't be doing any text matching. I'm fine with `mode: file` if you think that's cleaner, I don't feel strongly about this to be honest.

### Combining `pattern-file` clauses
Re: combining multiple `pattern-file`, I think we have a few options:
1. Intuitively, combining multiple `pattern-file` clauses in the same rule probably should be logically OR'd, as searching for files that match two file globs (logical AND) seems like it makes less sense to me when you'd want to do that.
2. Rather than implicitly doing logical OR and to give users more explicit flexibility, we could re-use `pattern-either` and `patterns` for if users want to OR or AND `pattern-file`s together.

If I had to choose, I think 2) is slightly nicer, as it's more consistent with other pattern clauses and makes Semgrep's behavior more explicit vs implicit. 

The following would match Python files whose name includes `bar` in it and ends in `.py` that reside in the `test/` directory. (logical AND between `pattern-file` clauses)

```yaml
patterns:
  - pattern-file: *bar*.py
  - pattern-file: test/*
```

The following would be how you'd do logical OR

```yaml
patterns:
  - pattern-either:
      - pattern-file: *bar*.py
      - pattern-file: test/*
```
### `pattern-not-file`
I'm also thinking a bit about a `pattern-not-file` clause, which would allow users to assert that they expect there to *not* be any file that matches a specific file path or glob. I'm not sure how useful this would be to people, so maybe we can wait on it.

### Priority: Low

A number of customers, and us internally, would find this useful. However, given the above mentioned workaround, this is a nice to have, not a fundamental blocker.

**Comment by IagoAbal** [comment]
I will then reopen the issue since it still seems relevant, not sure about its priority though (maybe it should be low?).

**Comment by aryx** [comment]
I think pattern-file could be combined with other patterns. We could make it match the range of the whole file,
so then it's just yet another range that can be combined with other. We could even have metavariable
in the glob like we have for metavariable-regex, metavariables that could be reused in other patterns or the message.

**Comment by aryx** [comment]
In fact, the paths field could be replaced by a series of pattern-file and pattern-either in theory.

**Comment by clintgibler** [comment]
Hm good point, `paths` could be replaced by `pattern-file` and `pattern-file-not`, which would both enable the workflow in this issue as well as replace the separate `paths` key. 

One downside is that it'll break existing rules that use `paths`, but I suppose we could continue supporting that internally for a few versions, and update the docs so people use the new way.

Thoughts from @dlukeomalley on `pattern-file` and `pattern-file-not` vs `paths`?

**Comment by Isaac0616** [comment]
A bit late to the discussion.
For the current workaround like:
```
rules:
  - id: found-dockerfile
    message: Detected a Dockerfile
    languages: [generic]
    severity: INFO
    pattern-regex: ^.*
    paths:
      include:
        - Dockerfile 
```
I think `^.*` would only match the first line.
Combining it with semgrep-action, the rule would only be triggered when the file is first added or the first line is modified.
Do I miss anything?
I was considering using `pattern-regex: ^` to match the empty string and handle the empty string specially in semgrep-action (or equivalent) so that any changes of the file would be flagged.
In any case, I would prefer either a special file mode or `pattern-file`/`pattern-file-not`.

**Comment by anuraagbaishya** [comment]
I was wondering if we can use this workaround to check if a file is missing? I have pondered over this for a while and could not think of a way to do this. 
My use case is I want to check if a JS repo does not have a `.npmrc` file. We have a guideline that requires repos to have a `.npmrc` file with some internal configurations. A missing file would indicate the configurations as a whole are missing for the repo
Alternatively - is it possible to make semgrep show an error if the path in a rule was not found? For eg if semgrep rule does not run in this path
```
  paths:
      include:
        - '.npmrc'
```
